requests.go

```diff
@@ -0,0 +1,276 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/cookiejar"
+	"strings"
+)
+
+const (
+	scheme                = "https://"
+	loginPagePath         = "/portal/"
+	loginClearPath        = "/portal/selfservice/users/loginClear/"
+	loginSubmitPath       = "/portal/login"
+	loginOauth2Path       = "/login/sessionCookieRedirect?"
+	oauth2RedirectUrlPath = "/myportal/"
+	userDataPath          = "/api/v1/sessions/me/lifecycle/refresh"
+	changePasswordPath    = "/myportal/viewprofile/myprofile/credential"
+	logoutPath            = "/myportal/logout"
+	userId                = "userId"
+	sessionToken          = "sessionToken"
+)
+
+type loginDetails struct {
+	Username string          `json:"username"`
+	Password string          `json:"password"`
+	Options  map[string]bool `json:"options"`
+}
+
+type changePassword struct {
+	OldPassword string `json:"oldPassword"`
+	NewPassword string `json:"newPassword"`
+}
+
+type UserData struct {
+	UserId       string `json:"userId"`
+	SessionToken string `json:"sessionToken"`
+}
+
+func sendRequest(client *http.Client, method, urlstr string, customHeaders map[string][]string, body []byte) (*UserData, error) {
+	var req *http.Request
+	var err error
+	userData := &UserData{}
+	if method == http.MethodGet {
+		req, err = http.NewRequest(http.MethodGet, urlstr, nil)
+	} else if method == http.MethodPost {
+		req, err = http.NewRequest(http.MethodPost, urlstr, bytes.NewReader(body))
+	} else {
+		return nil, errors.New("unsupported method type")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(body) > 0 {
+		req.Header.Add("Content-Type", contentType)
+	}
+
+	// Add common headers
+	req = addHeadersToRequest(req, commonHeaders)
+
+	// Add custom headers
+	req = addHeadersToRequest(req, customHeaders)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		b, _ := io.ReadAll(resp.Body)
+		return nil, errors.New(string(b))
+	}
+
+	if method == http.MethodPost && (strings.HasSuffix(urlstr, loginSubmitPath) ||
```

Comment:
Let me see if I understand. The sendRequest function takes an additional parameter: userData interface{}.
The code does this:
`	if userData != nil {
		err = json.NewDecoder(resp.Body).Decode(&userData)
		if err != nil {
			return err
		}
	}`
Since the function only returns `error`, doesn't the parameter userData have to be a pointer passed in by the caller?

Reply:
Done.